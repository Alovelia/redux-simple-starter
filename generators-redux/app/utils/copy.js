const fse = require('fs-extra');
const Promise = require('bluebird');
const fs = require('fs');
const path = require('path');
const curry = require('lodash/fp/curry');
const config = require('../config');
const _ = require('./mixins');
const warnings = require('./warnings');

// const lstatP = Promise.promisify(fs.lstat);
const chmod = Promise.promisify(fs.chmod);
let readFile = Promise.promisify(fs.readFile);

exports.copy = curry(async (src, dest) => {
  const srcResolved = path.resolve(config.TEMPLATE_ROOT, src || '');
  const destResolved = path.resolve(process.cwd(), dest || '');

  try {
    if (await fse.pathExists(destResolved)) {
      return warnings.showExistWarning(destResolved);
    }
    await fse.copy(srcResolved, destResolved, { errorOnExist: true });
    await chmod(destResolved, '755');
  } catch (err) {
    console.error(err);
  }
});

exports.copyTpl = curry(async (src, dest, data = {}) => {
  const srcResolved = path.resolve(config.TEMPLATE_ROOT, src || '');
  const destResolved = path.resolve(process.cwd(), dest || '');

  try {
    if (await fse.pathExists(destResolved)) {
      return warnings.showExistWarning(destResolved);
    }

    const tmplFile = await readFile(srcResolved);
    const compiled = _.template(tmplFile.toString())(Object.assign({}, _, data));
    await fse.outputFile(destResolved, compiled);
    await chmod(destResolved, '755');
  } catch (err) {
    console.error(err);
  }
});
