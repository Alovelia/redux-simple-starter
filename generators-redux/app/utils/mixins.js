let _ = require('lodash');
let humanize = require('humanize-string');

// TODO add tests for these functions

module.exports = _.mixin({
  camelCaseUpperFirst,
  dasherize,
  humanize,
  toConstantFormat,
  toConstantFormatRaw,
});

function camelCaseUpperFirst(value) {
  return _.flowRight(
    _.upperFirst,
    _.camelCase
  )(value);
}

function toConstantFormat(value) {
  return _.flowRight(
    _.toUpper,
    _.partialRight(_.replace, /-/g, '_'),
    _.kebabCase
  )(value);
}
// to use with ducks helpers ~ASYNC_ACTION
function toConstantFormatRaw(value) {
  const prefix = ~value.indexOf('~') ? '~' : '';
  return `${prefix}${toConstantFormat(value)}`;
}

// As far as I remember this function will make kabeb-case not lowUPPER -> low-u-p-p-e-r but low-upper
function dasherize(str) {
  let match = /[A-Z]+/.test(str),
    arr = str.split(''),
    copy = [],
    res = null,
    isUpperCase = function (letter) {
      return letter === letter.toUpperCase();
    };

  for (let i = 0, length = arr.length; i < length; i++) {
    copy[i] = arr[i];
  }

  for (let i = 0, length = arr.length; i < length; i++) {
    if (arr[i + 1] && isUpperCase(arr[i])) {
      res = isUpperCase(arr[i + 1]);
      if (res && arr[i - 1] && !isUpperCase(arr[i - 1])) {
        continue;
      }
      if (res || res === false && i === 0) {
        copy[i] = arr[i].toLowerCase();
      }
    } else if (!arr[i + 1]) {
      copy[i] = arr[i].toLowerCase();
    }
  }
  return copy.join('')
    .replace(/([A-Z]+)/g, '-$1')
    .replace(/[-_\s]+/g, '-')
    .toLowerCase();
}
