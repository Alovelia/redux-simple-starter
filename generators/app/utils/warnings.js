let chalk = require('chalk');

const log = console.log.bind(console);

exports.showWarning = function showWarning(text) {
  log(chalk.green('--------Info--------'));
  log(chalk.yellow('You did not have this folder!'));
  log(chalk.yellow(text));
  log(chalk.yellow('I have created folder structure for you'));
  log(chalk.green('-----------------------'));
};

exports.showExistWarning = function showExistWarning(fName) {
  log(chalk.yellow('--------Warning--------'));
  log(chalk.yellow('You have this file'));
  log(chalk.red(fName));
  log(chalk.yellow('skip...'));
  log(chalk.yellow('-----------------------'));
};
