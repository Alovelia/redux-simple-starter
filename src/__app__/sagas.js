import { delay } from 'redux-saga';
import { all, call, fork } from 'redux-saga/effects';
// #if process.env.NODE_ENV === 'development'
import homeSagas from '../home/sagas';
//†import
// #endif

// https://github.com/redux-saga/redux-saga/issues/760#issuecomment-273737022
const makeRestartable = (saga) => {
  return function* sagaRunner() {
    // yield spawn(function* restartableSaga() {
    yield fork(function* restartableSaga() {
      while (true) {
        try {
          yield call(saga);
          // eslint-disable-next-line
          console.error('unexpected root saga termination. ' +
            'The root sagas are supposed to be sagas that live during the whole app lifetime!', saga);
        } catch (e) {
          // eslint-disable-next-line
          console.error('Saga error, the saga will be restarted', e);
        }
        yield delay(1000); // Avoid infinite failures blocking app TODO use backoff retry policy...
      }
    });
  };
};


const rootSagas = [
  // #if process.env.NODE_ENV === 'development'
  // to enable hot module reload sagas should be combined in one place as single process
  // only for development
  ...homeSagas,
  //†saga
  // #endif
].map(makeRestartable);


export default function* root() {
  // https://github.com/redux-saga/redux-saga/issues/1000#issuecomment-303760253
  // https://github.com/redux-saga/redux-saga/issues/171#issuecomment-345042076
  yield all([
    ...rootSagas.map(saga => call(saga))
  ]);
}
