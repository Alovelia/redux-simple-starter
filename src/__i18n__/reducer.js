import _ from 'common/helpers';
import { constants, actions } from 'ducks-helpers';
import handleActions from 'redux-actions/lib/handleActions';
import { getDefaultSupportedLanguage } from './core';

export const TYPE = constants('@@intl', [
  'UPDATE_LOCALE',
  'UPDATE_MESSAGES',
  //†type
]);

export const ACTION = actions(TYPE);

export const initialState = _.fromJS({
  locale: getDefaultSupportedLanguage(),
  messages: {},
});

export default handleActions({
  [TYPE.UPDATE_LOCALE]: updateLocale,
  [TYPE.UPDATE_MESSAGES]: updateMessages,
  //†handler
}, initialState);


export function updateLocale(state, { payload }) {
  return state.set('locale', payload);
}
export function updateMessages(state, { payload }) {
  return state.set('messages', _.fromJS(payload));
}
//†reducer
