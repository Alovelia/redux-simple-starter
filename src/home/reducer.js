import { constants, actions } from 'ducks-helpers';
import handleActions from 'redux-actions/lib/handleActions';
import _ from 'common/helpers';

export const TYPE = constants('home', [
  '~GET',
  //†type
]);
export const ACTION = actions(TYPE);

// The initial state
export const initialState = _.fromJS({});

export default handleActions({
  [TYPE.GET_SUCCESS]: getSuccess,
  //†handler
}, initialState);

function getSuccess(state) {
  return state.set('username', 'username');
}
//†reducer
