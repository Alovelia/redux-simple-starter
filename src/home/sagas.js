import { call, put, select, takeLatest } from 'redux-saga/effects';
// import { LOCATION_CHANGE } from 'react-router-redux';
import axios from 'axios';
import { ACTION, TYPE } from './reducer';
import { makeSelectUsername } from './selectors';

export function* getData() {
  // Select username from store
  const username = yield select(makeSelectUsername());

  try {
    // Call ajax
    // eslint-disable-next-line
    const response = yield call(axios, '/api/mock');
    yield put(ACTION.getSuccess(username));
    yield put({ type: 'SOMETHING_ELSE' });
  } catch (err) {
    yield put(ACTION.getError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* getDataSaga() {
  // Watches for action and calls callback when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // const watcher = yield takeLatest(TYPE.GET, getData);
  yield takeLatest(TYPE.GET, getData);

  // Suspend execution until location changes
  // yield take(LOCATION_CHANGE);
  // yield cancel(watcher);
}

// Bootstrap sagas
export default [
  getDataSaga,
];
