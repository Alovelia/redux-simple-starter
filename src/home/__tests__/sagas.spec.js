import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';
import { LOCATION_CHANGE } from 'react-router-redux';
import { ACTION, TYPE } from '../reducer';

import { getData, getDataSaga } from '../sagas';

describe('getData Saga', () => {
  let getDataGenerator;
  const username = 'troll';
  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getDataGenerator = getData();

    const selectDescriptor = getDataGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getDataGenerator.next(username).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the getSuccess action if it requests the data successfully', () => {
    const response = {
      username
    };
    const putDescriptor = getDataGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(ACTION.getSuccess(username)));
  });


  it('should dispatch SOMETHING_ELSE if it requests the data successfully', () => {
    const somethingElseDescriptor = getDataGenerator.next().value;
    expect(somethingElseDescriptor).toMatchSnapshot();
  });

  it('should call the getError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getDataGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(ACTION.getError(response)));
  });
});

describe('getDataSaga Saga', () => {
  const getDataSagaGenerator = getDataSaga();
  const mockedTask = createMockTask();

  it(`should start task to watch for ${TYPE.GET} action`, () => {
    const takeLatestDescriptor = getDataSagaGenerator.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(TYPE.GET, getData));
  });
});
