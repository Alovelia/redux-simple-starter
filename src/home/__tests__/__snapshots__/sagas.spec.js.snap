// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRepos Saga should call the getError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getRepos Saga should call the getError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/mock",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getRepos Saga should dispatch SOMETHING_ELSE if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getRepos Saga should dispatch SOMETHING_ELSE if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/mock",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getRepos Saga should dispatch SOMETHING_ELSE if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "meta": undefined,
      "payload": "troll",
      "type": "home/GET_SUCCESS",
    },
    "channel": null,
  },
}
`;

exports[`getRepos Saga should dispatch the getSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getRepos Saga should dispatch the getSuccess action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/mock",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
